lib/src/model/analysis/retro_controller.dart

```diff
@@ -0,0 +1,537 @@
+      // is good enough to be considered an alternative solution.
+      ? ref
+            .read(engineEvaluationPreferencesProvider)
+            .copyWith(engineSearchTime: kMaxEngineSearchTime)
+      : ref.read(engineEvaluationPreferencesProvider);
+
+  @override
+  @protected
+  EngineEvaluationPreferences get evaluationPreferencesNotifier =>
+      ref.read(engineEvaluationPreferencesProvider.notifier);
+
+  @override
+  @protected
+  EvaluationService evaluationServiceFactory() => ref.read(evaluationServiceProvider);
+
+  @override
+  @protected
+  EvaluationMixinState get evaluationState => state.requireValue;
+
+  @override
+  @protected
+  late SocketClient socketClient;
+
+  @override
+  @protected
+  Node get positionTree => _root;
+
+  @override
+  Future<RetroState> build(RetroOptions options) async {
+    ref.onDispose(() {
+      disposeEngineEvaluation();
+    });
+
+    socketClient = ref.watch(socketPoolProvider).open(AnalysisController.socketUri);
+
+    _game = await ref.read(archivedGameProvider(id: options.id).future);
+    _root = _game.makeTree();
+
+    initEngineEvaluation();
+
+    // We need to define the state value in the build method because `requestEval` require the state
+    // to have a value.
+    state = AsyncData(await _initState(options.initialSide));
+
+    socketClient.firstConnection.then((_) {
+      requestEval();
+    });
+
+    return state.requireValue;
+  }
+
+  Future<RetroState> _initState(Side side) async {
+    final mistakes = (await Future.wait(
+      _root.view.mainline.map((branch) async {
+        if (branch.position.turn != side || branch.children.isEmpty) {
+          return null;
+        }
+
+        final eval = branch.serverEval;
+        final newEval = branch.children.first.serverEval;
+
+        if (eval == null || newEval == null) {
+          return null;
+        }
+
+        // https://github.com/lichess-org/lila/blob/d29f27d8cbb0e0dac38308c23c63b828028c085f/ui/analyse/src/nodeFinder.ts#L26
+        final bigEvalSwing = Eval.winningChancesPovDiff(side, eval, newEval).abs() > 0.1;
+
+        final lostEasyMate = eval.mate != null && newEval.mate == null && eval.mate!.abs() <= 3;
+
+        final hasSolution = branch.children.length > 1;
+
+        final isMistake = (bigEvalSwing || lostEasyMate) && hasSolution;
+        if (!isMistake) return null;
+
+        var openingExplorerSolutions = const IList<UCIMove>.empty();
+        final middlegame = _game.meta.division?.middlegame;
+        if (middlegame == null || branch.position.ply + 1 < middlegame) {
+          try {
+            final entry = await ref
+                .read(openingExplorerRepositoryProvider)
+                .getMasterDatabase(branch.position.fen, since: MasterDb.kEarliestYear);
+
+            // https://github.com/lichess-org/lila/blob/d28f27d8cbb0e0dac38308c23c63b828028c085f/ui/analyse/src/retrospect/retroCtrl.ts#L108
```

Comment:
I think we can remove this comment; this is sufficiently self explanatory. The link will most probably be out of date at some point so better not having it.